doc/man/man3/post_upload_actions.3
.TH "post_upload_actions" 3 "Fri Sep 17 2021" "Version 1.0" "DIY Auto-Correlator" \" -*- nroff -*-
.ad l
.nh
.SH NAME
post_upload_actions
.SH SYNOPSIS
.br
.PP
.SS "Functions"

.in +1c
.ti -1c
.RI "def \fBupdate_fn\fP ()"
.br
.in -1c
.SS "Variables"

.in +1c
.ti -1c
.RI "\fBconfig\fP = json\&.load(f)"
.br
.ti -1c
.RI "\fBx_tau_values\fP = x_tics(\fBconfig\fP['LinCorrs'], \fBconfig\fP['SeriesSize'], \fBconfig\fP['BinRatio'])"
.br
.ti -1c
.RI "\fBnorm_values\fP = points_norm_template(\fBconfig\fP['LinCorrs'], \fBconfig\fP['SeriesSize'], \fBconfig\fP['BinRatio'])"
.br
.ti -1c
.RI "\fBname_\fP = str(time\&.time_ns())"
.br
.ti -1c
.RI "string \fBy_file_name\fP = f'{name_}_y\&.dat'"
.br
.ti -1c
.RI "string \fBx_file_name\fP = f'{name_}_x_taus\&.dat'"
.br
.ti -1c
.RI "\fBfmt\fP"
.br
.ti -1c
.RI "\fBnewline\fP"
.br
.ti -1c
.RI "\fBdatafile\fP = open(\fBy_file_name\fP, 'a')"
.br
.ti -1c
.RI "\fBport\fP = Serial(port=\fBconfig\fP['Port'], baudrate=\fBconfig\fP['Baud'], timeout=0\&.5)"
.br
.ti -1c
.RI "\fBapp\fP = QtGui\&.QApplication([])"
.br
.ti -1c
.RI "\fBwindow\fP = pg\&.GraphicsLayoutWidget(show=True, title= f'Auto\-Correlation Plot \- {\fBconfig\fP['Port']}')"
.br
.ti -1c
.RI "\fBantialias\fP"
.br
.ti -1c
.RI "\fBgraph\fP = window\&.addPlot(title='Auto\-Correlation Plot')"
.br
.ti -1c
.RI "\fBcurve\fP = graph\&.plot(pen='y')"
.br
.ti -1c
.RI "\fBwait_time_ms\fP = \fBconfig\fP['SamplingDelay_ms']"
.br
.ti -1c
.RI "\fBtimer\fP = QtCore\&.QTimer()"
.br
.in -1c
.SH "Function Documentation"
.PP 
.SS "def post_upload_actions\&.update_fn ()"

.SH "Variable Documentation"
.PP 
.SS "post_upload_actions\&.antialias"

.SS "post_upload_actions\&.app = QtGui\&.QApplication([])"

.SS "post_upload_actions\&.config = json\&.load(f)"

.SS "post_upload_actions\&.curve = graph\&.plot(pen='y')"

.SS "post_upload_actions\&.datafile = open(\fBy_file_name\fP, 'a')"

.SS "post_upload_actions\&.fmt"

.SS "post_upload_actions\&.graph = window\&.addPlot(title='Auto\-Correlation Plot')"

.SS "post_upload_actions\&.name_ = str(time\&.time_ns())"

.SS "post_upload_actions\&.newline"

.SS "post_upload_actions\&.norm_values = points_norm_template(\fBconfig\fP['LinCorrs'], \fBconfig\fP['SeriesSize'], \fBconfig\fP['BinRatio'])"

.SS "post_upload_actions\&.port = Serial(port=\fBconfig\fP['Port'], baudrate=\fBconfig\fP['Baud'], timeout=0\&.5)"

.SS "post_upload_actions\&.timer = QtCore\&.QTimer()"

.SS "int post_upload_actions\&.wait_time_ms = \fBconfig\fP['SamplingDelay_ms']"

.SS "post_upload_actions\&.window = pg\&.GraphicsLayoutWidget(show=True, title= f'Auto\-Correlation Plot \- {\fBconfig\fP['Port']}')"

.SS "string post_upload_actions\&.x_file_name = f'{name_}_x_taus\&.dat'"

.SS "post_upload_actions\&.x_tau_values = x_tics(\fBconfig\fP['LinCorrs'], \fBconfig\fP['SeriesSize'], \fBconfig\fP['BinRatio'])"

.SS "string post_upload_actions\&.y_file_name = f'{name_}_y\&.dat'"

.SH "Author"
.PP 
Generated automatically by Doxygen for DIY Auto-Correlator from the source code\&.


code/python/post_upload_actions.py
import time
import threading
import json
import numpy as np
import math
import serial.tools.list_ports as port_list
from serial import Serial, SerialException

from multitau import x_tics, points_norm_template, channel_size
from utilities import PseudoConsole, WelcomeText, periodic_callbacks



WelcomeText()


# Load Configurarion file
with open('./../../config.json', 'r') as f:
	config = json.load(f)
channel_size = multitau.channel_size(config['LinCorrs'], config['SeriesSize'], config['BinRatio'])
byte_size = 4 # No of bytes used by the incoming data for one value

x_tau_values = x_tics(config['LinCorrs'], config['SeriesSize'], config['BinRatio'])

#Generate Normalization
norm_values = points_norm_template(config['LinCorrs'], config['SeriesSize'], config['BinRatio'])


 
#Set Files
name_ = str(time.time_ns())
y_file_name = f"{name_}_y.dat"
x_file_name = f"{name_}_x_taus.dat"
print(f"Saving data to: ** [x: {x_file_name} , y: {y_file_name}]  **")
with open(x_file_name, 'w') as f:
	np.savetxt(f, x_tau_values.flatten(), fmt='%s', newline = ", ")
datafile = open(y_file_name, 'a')

# Set Live Graph if enabled
if config["LiveGraph"] == True:
	#Create Receiver object to the com port

	from pyqtgraph.Qt import QtGui, QtCore
	import pyqtgraph as pg
	
	app = QtGui.QApplication([])
	
	
	window = pg.GraphicsLayoutWidget(show=True, title= f"Auto-Correlation Plot - {config['Port']}")
	window.resize(1440,720)
	window.setWindowTitle("Auto-Correlation Plot")


	# icon
	icon = QtGui.QIcon("skin.jpg")
	window.setWindowIcon(icon)


	# Enable antialiasing for prettier plots
	pg.setConfigOptions(antialias=True, useOpenGL=True)

	canvas = window.addPlot(title="Auto-Correlation Plot → 0")
	window.setCentralWidget(canvas)

	

	x_right_padding = math.pow(10, math.floor(math.log(x_tau_values[-1], 10))/2)
	#canvas.setXRange(0, x_tau_values[-1])# + x_right_padding)

	#canvas.scaleBy(s=None, center=None, x=None, y=None)
	curve = canvas.plot(pen='y', symbolBrush='c', symbolSize=4)
	curve.setLogMode(False, False)
	canvas.showGrid(x = True, y = True)
	canvas.setLabel('left', 'ACF', 'G(Tau)')
	canvas.setLabel('bottom', 'Lag', 'Tau')
	canvas.setRange(xRange=(0, x_tau_values[-1]), yRange=None, padding=None, update=True, disableAutoRange=True)



# Open Serial Port
port = Serial(port=config["Port"], baudrate=config["Baud"], timeout=None)
if not port.is_open:
	raise Exception("Port could not be opened.")
	

#Defines update function
update_id = 0
time_init = time.perf_counter()
time_stop = time.perf_counter()
def update_fn():
	if port.in_waiting >= channel_size * byte_size:
		raw_data = port.read(channel_size * byte_size)
		data = np.frombuffer(raw_data, dtype = np.uint32, count = channel_size)
		global update_id
		update_id = update_id + 1
		# Points is the number of points read by LC0
		# raw_data = port.readline()
		# no_points = np.frombuffer(raw_data, dtype = '<i4')
		# data = np.frombuffer(raw_data, dtype = '<i4', offset = 4)
		# current_norm  = norm_values * np.unsignedinteger(no_points)
		normed_y = data #/ norm_values
		np.savetxt(datafile, normed_y.flatten(), fmt='%s', newline = ", ")
		datafile.write('\n')

		if config["LiveGraph"] == True:
			global time_init, time_stop
			time_stop = time.perf_counter()
			canvas.setTitle(title=f"Auto-Correlation Plot → {update_id} - Update time: {(time_stop - time_init): .3f} s")
			curve.setData(x_tau_values, normed_y)
			time_init = time.perf_counter()


# Set Sampling Rate
if 'SamplingDelay_ms' in config:
	wait_time_ms = config['SamplingDelay_ms']
else:
	wait_time_ms = 50 

print(f"""Sampling Time: {wait_time_ms}\nPort: {config['Port']}""")


stop_code_asserted = False
# Thread Division
if config["LiveGraph"] == True:
	print("••• Launching Live Graph → Close graph window to end measurement.")
	timer = QtCore.QTimer()
	timer.timeout.connect(update_fn)
	timer.start(wait_time_ms)
	app.exec_()
else: 
	
	def periodic_callback_fn(period_ms=wait_time_ms):
		period_s = period_ms * 0.001 #Convert to seconds
		def g_tick():
			t = time.time()
			while True:
				t += period_s
				yield max(t - time.time(), 0) #Avoid negative values
		sleep_time_gen = g_tick()
		global stop_code_asserted
		while not stop_code_asserted:
			time.sleep(next(sleep_time_gen))
			update_fn()
	
	timer = threading.Thread(target=periodic_callback_fn)
	timer.start()
	print(f"Measurement Thread -> {timer.is_alive()*'Alive' + (not timer.is_alive()) * 'Dead' }")
	PseudoConsole()




# Stop Procedure
if config["LiveGraph"] == True:
	timer.stop()
else:
	stop_code_asserted = True
	print(f"Waiting for measurement thread to join...")
	timer.join()

datafile.close()
port.close()
print(f"•••> Exiting: Updates read: {update_id}")

